File: source_formatting/shulchan_arukh_remove_header.py
#!/usr/bin/python

from source_formatting.html_parser import BaseHtmlTranslator

class ShulchanArukhHeaderRemover(BaseHtmlTranslator):
    tag_stack = []
    processing_header = False
    done_processing_header = False

    def handle_starttag(self, tag, attrs):
        if self.done_processing_header and tag == "br" and len(self.tag_stack) == 0:
            return

        self.tag_stack.append(tag)

        if not self.done_processing_header and tag == "b" and self.tag_stack == ["b"]:
            self.processing_header = True
            return

        if self.processing_header:
            return

        self.append_start_tag(tag, attrs)

    def handle_endtag(self, tag):
        self.tag_stack.pop()

        if not self.done_processing_header and tag == "b" and len(self.tag_stack) == 0:
            self.done_processing_header = True
            self.processing_header = False
            return

        if self.processing_header:
            return

        self.append_end_tag(tag)

    def handle_data(self, data):
        if self.processing_header:
            return
        self._out.append(data)

File: source_formatting/shulchan_arukh_remove_header.ts
import {Attributes, HtmlVisitor} from "./html_visitor";

export class ShulchanArukhHeaderRemover extends HtmlVisitor {
  tagStack: string[] = []
  processingHeader = false
  doneProcessingHeader = false

  visitStartTag(tag: string, attributes: Attributes): void {
    if (this.doneProcessingHeader && tag === "br" && this.tagStack.length === 0) {
      return;
    }

    this.tagStack.push(tag);
    if (!this.doneProcessingHeader && tag === "b" && this.tagStack.length === 1 && this.tagStack[0] === "b") {
      this.processingHeader = true;
      return;
    }

    if (this.processingHeader) {
      return;
    }

    this.appendStartTag(tag, attributes);
  }

  visitEndTag(tag: string): void {
    this.tagStack.pop();

    if (!this.doneProcessingHeader && tag === "b" && this.tagStack.length === 0) {
      this.doneProcessingHeader = true;
      this.processingHeader = false;
      return;
    }

    if (this.processingHeader) {
      return;
    }

    this.appendEndTag(tag);
  }

  visitText(text: string): void {
    if (!this.processingHeader) {
      this.appendText(text);
    }
  }
}
